Update error notifications guide (#216)
astronomer/airflow-guides: guides/error-notifications-in-airflow.md

---
title: "Error Notifications in Airflow"
description: "Methods for managing notifications in your Airflow DAGs."
date: 2018-05-21T00:00:00.000Z
slug: "error-notifications-in-airflow"
heroImagePath: null
tags: ["DAGs", "Integrations", "Operators"]
---

## Overview

> A key question when using any data orchestration tool is "How do I know if something has gone wrong?". Airflow users always have the option to check the UI to see the status of their DAGs, but this is an inefficient way of managing errors systematically, especially if certain failures need to be addressed promptly or by multiple team members. Fortunately, Airflow has built-in notification mechanisms that can be leveraged to configure error notifications in a way that works for your team. 

Reviewer: ```suggestion
A key question when using any data orchestration tool is "How do I know if something has gone wrong?". Airflow users can check the UI to see the status of their DAGs, but this is an inefficient way of managing errors systematically, especially if certain failures need to be addressed promptly or by multiple team members. The better option is to use Airflow's built-in notification system to programmatically alert your team about errors.
```

        slack_msg = """
                :red_circle: Task Failed. 
                *Task*: {task}  
                *Dag*: {dag} 
                *Execution Time*: {exec_date}  
                *Log Url*: {log_url} 
                """.format(
                task=context.get('task_instance').task_id,
                dag=context.get('task_instance').dag_id,
                ti=context.get('task_instance'),
                exec_date=context.get('execution_date'),
                log_url=context.get('task_instance').log_url,
            )
        failed_alert = SlackWebhookOperator(
            task_id='slack_notification',
            http_conn_id='slack_webhook',
            message=slack_msg)
        return failed_alert.execute(context=context)
    ```

    > Note: in Airflow 2.0 or greater, to use the `SlackWebhookOperator` you will need to install the `apache-airflow-providers-slack` provider package.

 4. Define your `on_failure_callback` parameter in your DAG either as a `default_arg` for the whole DAG, or for specific tasks. Set it equal to the function you created in the previous step. You should now see any failure notifications show up in Slack!

## Airflow SLAs

> [Airflow SLAs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#slas) are a type of notification that you can leverage in Airflow if your tasks are taking longer than expected to complete. SLAs are defined as the maximum amount of time the task should take; if the task takes longer than this amount of time, the SLA will be missed and notifications will be triggered. This may be useful in cases where you have potentially long-running tasks that might require user intervention after a certain period of time, or if you have tasks that need to complete by a certain deadline. Note that exceeding an SLA will not stop a task from running; if this is the desired behavior, check out [timeouts](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#timeouts) instead.

Reviewer: ```suggestion
[Airflow SLAs](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#slas) are a type of notification that you can use if your tasks are taking longer than expected to complete. If a task takes longer than a maximum amount of time to complete as defined in the SLA, the SLA will be missed and notifications will be triggered. This can be useful in cases where you have potentially long-running tasks that might require user intervention after a certain period of time, or if you have tasks that need to complete by a certain deadline. 

Note that exceeding an SLA will not stop a task from running. If you want tasks to stop running after a certain time, try using [timeouts](https://airflow.apache.org/docs/apache-airflow/stable/concepts/tasks.html#timeouts) instead.
```

```python
from airflow import DAG
from airflow.operators.dummy_operator import DummyOperator
from airflow.operators.python_operator import PythonOperator
from datetime import datetime, timedelta
import time

def my_custom_function(ts,**kwargs):
    print("task is sleeping")
    time.sleep(40)

# Default settings applied to all tasks
default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email_on_failure': True,
    'email': 'kenten@astronomer.io',
    'email_on_retry': False,
    'sla': timedelta(seconds=30)
}

# Using a DAG context manager, you don't have to specify the dag property of each task
with DAG('sla-dag',
         start_date=datetime(2021, 1, 1),
         max_active_runs=1,
         schedule_interval=timedelta(minutes=2),
         default_args=default_args,
         catchup=False 
         ) as dag:

    t0 = DummyOperator(
        task_id='start'
    )

    t1 = DummyOperator(
        task_id='end'
    )

    sla_task = PythonOperator(
        task_id='sla_task',
        python_callable=my_custom_function
    )

    t0 >> sla_task >> t1
```

> SLAs can be a bit unintuitive; here are some notes to help understand SLA behavior:

Reviewer: ```suggestion
SLAs have some unique behaviors that you should consider before implementing them:
```

> - SLAs are relative to the DAG execution date, not the task start time. For example, in the DAG above the `sla_task` will miss the 30 second SLA because it takes at least 40 seconds to complete. The `t1` task will also miss the SLA, because it is executed more than 30 seconds after the DAG execution date. In that case the `sla_task` will be considered "blocking" to the `t1` task.

Reviewer: ```suggestion
- SLAs are relative to the DAG execution date, not the task start time. For example, in the DAG above the `sla_task` will miss the 30 second SLA because it takes at least 40 seconds to complete. The `t1` task will also miss the SLA, because it is executed more than 30 seconds after the DAG execution date. In that case, the `sla_task` will be considered "blocking" to the `t1` task.
```

- SLAs will only be evaluated on scheduled DAG Runs. They will not be evaluated on manually triggered DAG Runs.

> - SLAs can be set at the task level if a different SLA is required for each task. In this case, all task SLAs are still relative to the DAG execution date. For example, in the DAG below, `t1` has an SLA of 500 seconds. If the upstream tasks (`t0` and `sla_task`) combined take 450 seconds to complete, and `t1` takes 60 seconds to complete, `t1` will miss its SLA even though the task did not take more than 500 seconds to execute.

Reviewer: ```suggestion
- SLAs can be set at the task level if a different SLA is required for each task. In this case, all task SLAs are still relative to the DAG execution date. For example, in the DAG below, `t1` has an SLA of 500 seconds. If the upstream tasks (`t0` and `sla_task`) combined take 450 seconds to complete, and `t1` takes 60 seconds to complete, then `t1` will miss its SLA even though the task did not take more than 500 seconds to execute.
```

```python
from airflow import DAG
from airflow.operators.dummy_operator import DummyOperator
from airflow.operators.python_operator import PythonOperator
from datetime import datetime, timedelta
import time

def my_custom_function(ts,**kwargs):
    print("task is sleeping")
    time.sleep(40)

# Default settings applied to all tasks
default_args = {
    'owner': 'airflow',
    'depends_on_past': False,
    'email_on_failure': True,
    'email': 'kenten@astronomer.io',
    'email_on_retry': False
}

# Using a DAG context manager, you don't have to specify the dag property of each task
with DAG('sla-dag',
         start_date=datetime(2021, 1, 1),
         max_active_runs=1,
         schedule_interval=timedelta(minutes=2),
         default_args=default_args,
         catchup=False 
         ) as dag:

    t0 = DummyOperator(
        task_id='start',
        sla=timedelta(seconds=50)
    )

    t1 = DummyOperator(
        task_id='end',
        sla=timedelta(seconds=500)
    )

    sla_task = PythonOperator(
        task_id='sla_task',
        python_callable=my_custom_function,
        sla=timedelta(seconds=5)
    )

    t0 >> sla_task >> t1
```

> Any SLA misses will be shown in the Airflow UI; you can view them by going to Browse → SLA Misses, which will look something like this:

Reviewer: ```suggestion
Any SLA misses will be shown in the Airflow UI. You can view them by going to Browse → SLA Misses, which looks something like this:
```

![SLA UI](https://assets2.astronomer.io/main/guides/error-notifications/sla_ui_view.png)

> If you have an SMTP server configured in your Airflow environment, you will also receive an email with notifications of any missed SLAs.

Reviewer: ```suggestion
If you configured an SMTP server in your Airflow environment, you will also receive an email with notifications of any missed SLAs.
```
